### ISSUE TITLE
Tweet.Length working in UTF-16, Twitter works in UTF-32

### ISSUE THREAD

**Issue by JoshKeegan** [linvi/tweetinvi#88]
If you call Tweet.Length(), the resulting length is given for the UTF-16 representation of the string (as is used in C#), but the character count for a Tweet gets enforced by Twitter as 140 actual characters (i.e. in UTF-32). So any characters that take more than 2 bytes to represent get counted as 2 characters, even though Twitter will only count them as 1.

So, if you put the following text in Tweet.Length(), you will get 141 characters, but if you post it to Twitter it will see it as 140 and allow for it to be posted:
```
Test full length status with a unicode char 🚒 that takes up 4 bytes (2 units in UTF-16) jhasdbcjkasdhfkjasdhfkjahjkfjasdhjfkjashfkjasdh end1
```
This is because of the 🚒 character in the text.

Whether or not you change the behaviour of Tweet.Length() I think this probably needs documenting carefully, as if undocumented it could be quite confusing as to why c# string.Length returns something different to Tweetinvi Tweet.Length() without any media or URLs.

This isn't something I use, so isn't particularly important to me, but I'm currently going through my own code fixing the same problem and noticed it was in Tweetinvi too :)

**Comment by linvi** [comment]
Thanks again Josh for reporting all these issues :)

**Comment by linvi** [comment]
Hi, have you find a good solution to detect UTF-32 characters? Currently I have find a solution that rely on error handling. Which I would really like to avoid.

**Comment by JoshKeegan** [comment]
Hi Linvi,
You can iterate over each character in the string looking for surrogate pairs (two UTF-16 chars being used to represent a single actual character). When you find a surrogate pair, don't count the second character towards the length. 
You can identify a surrogate pair because the chars will lie within reserved ranges.

Have a look at https://github.com/JoshKeegan/HtmlEncoder/blob/master/HtmlEncoder/StringIndicesConverter.cs which converts indices for the UTF-32 representation of a string to what they will be in the UTF-16 representation. Whilst this isn't exactly what you want, the detection of a surrogate pair is, and correcting the length is trivial if you can find surrogate pairs.
Cheers,
Josh

**Comment by linvi** [comment]
Thanks for your help. This will be available in version 0.9.10.1.
